Building Firmware from Git Commits

# Building Firmware from Git Commits

Timeline: End of October

## Task

Create a script that, given a Git commit hash, checks out the `racecar/` repository at that commit and builds a project for the SIL ("raspi") platform.

Eventually, this script will be written in Go, but you can prototype it in any language.

After trying to build a SIL project on Windows, I don't think it's possible. I was able to setup gRPC successfully on windows, but the issue is the CAN dependencies which rely on Linux specific headers. There may be more, but I can confirm that the following headers in `racecar\firmware\mcal\raspi\periph\can.h` induce an error:
```c
#include <linux/can.h>
#include <linux/can/raw.h>
```
There doesn't seem to be a direct replacement for Windows compilation, and we probably would not want it if we're building for a Raspberry Pi ([source](https://stackoverflow.com/questions/45477664/linux-includes-in-windows-ioctl-etc)).

There are two potential solutions I can think of:

1. Using the [arm-none-linux-gnueabihf](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads) toolchain might allow a Windows system to build a project with Linux dependencies. This may interfere with the bare metal target toolchain and might require changes to the build process, as well as adding to the dev environment setup for HIL/SIL.
2. Simply downloading the required Linux headers and trying to link it manually, maybe by adding it to the included header files contained within the bare metal toolchain. This probably won't work.

I can look into these later, but for now I will switch to an Ubuntu VM.
The exact make command and output is shown below, including the fatal error:
<details><summary>Console</summary>
<p>

```zsh
$ make PROJECT=Demo/CAN/Bar PLATFORM=raspi build
cmake -B build/Demo/CAN/Bar/raspi -S. -G"Unix Makefiles" -D PROJECT=Demo/CAN/Bar -D PLATFORM=raspi -D PROJECT_NAME=Bar --log-level STATUS
Build type: Debug
-- Validating input arguments
--   Verifying Demo/CAN/Bar is a project
--   The directory "C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/projects/Demo/CAN/Bar" exists.
--   Verifying Demo/CAN/Bar is a project - Valid
--   Verifying raspi is a platform under Demo/CAN/Bar
--   The directory "C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/projects/Demo/CAN/Bar/platforms/raspi" exists.
--   Verifying raspi is a platform under Demo/CAN/Bar - Valid
-- Validating input arguments - Success
-- Validating the MCAL Platform
--   Searching for mcal configuration file
--   The file "C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/projects/Demo/CAN/Bar/platforms/raspi/mcal_conf.cmake" exists.
--   Searching for mcal configuration file - Found
--   Verifying the raspi MCAL exists
--   The directory "C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/mcal/raspi" exists.
--   Verifying the raspi MCAL exists - Valid MCAL
--   Finding MCAL postbuild file
--   The file "C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/mcal/raspi/PostBuild.cmake" exists.
--   Finding MCAL postbuild file - Found
--   Finding MCAL toolchain file
--   The file "C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/mcal/raspi/Toolchain.cmake" exists.
--   Finding MCAL toolchain file - Found
-- Validating the MCAL Platform - Success
-- Building "Demo/CAN/Bar" for "raspi" with mcal "raspi".
CMake Warning (dev) at CMakeLists.txt:108 (enable_language):
  project() should be called prior to this enable_language() call.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at C:/tools/msys64/mingw64/share/cmake/Modules/Platform/Windows-GNU.cmake:193 (enable_language):
  project() should be called prior to this enable_language() call.
Call Stack (most recent call first):
  C:/tools/msys64/mingw64/share/cmake/Modules/Platform/Windows-GNU-C.cmake:2 (__windows_compiler_gnu)
  C:/tools/msys64/mingw64/share/cmake/Modules/CMakeCInformation.cmake:48 (include)
  CMakeLists.txt:108 (enable_language)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at C:/tools/msys64/mingw64/share/cmake/Modules/Platform/Windows-GNU.cmake:193 (enable_language):
  project() should be called prior to this enable_language() call.
Call Stack (most recent call first):
  C:/tools/msys64/mingw64/share/cmake/Modules/Platform/Windows-GNU-CXX.cmake:2 (__windows_compiler_gnu)
  C:/tools/msys64/mingw64/share/cmake/Modules/CMakeCXXInformation.cmake:48 (include)
  CMakeLists.txt:108 (enable_language)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at C:/tools/msys64/mingw64/share/cmake/Modules/Platform/Windows-GNU.cmake:193 (enable_language):
  project() should be called prior to this enable_language() call.
Call Stack (most recent call first):
  C:/tools/msys64/mingw64/share/cmake/Modules/Platform/Windows-GNU-ASM.cmake:2 (__windows_compiler_gnu)
  C:/tools/msys64/mingw64/share/cmake/Modules/CMakeASMInformation.cmake:38 (include)
  CMakeLists.txt:108 (enable_language)
This warning is for project developers.  Use -Wno-dev to suppress it.

C:/Python311/python.exe
-- etl | Version string determined with git describe: 20.38.10-2-g84031659f
-- etl | Determined ETL version 20.38.10 from the git tag
-- Configuring done (0.6s)
-- Generating done (0.2s)
-- Build files have been written to: C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/build/Demo/CAN/Bar/raspi
Copying compile_commands.json to build for clangd.
cp build/Demo/CAN/Bar/raspi/compile_commands.json build/compile_commands.json
Ensure the .clangd file exists.
touch .clangd 
cmake --build build/Demo/CAN/Bar/raspi
make[1]: Entering directory 'C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/build/Demo/CAN/Bar/raspi'
[  9%] Running cpp protocol buffer compiler on signals.proto
[ 18%] Running grpc protocol buffer compiler on signals.proto
[ 27%] Building CXX object validation/sil/CMakeFiles/sil.dir/signals.pb.cc.obj
[ 36%] Building CXX object validation/sil/CMakeFiles/sil.dir/signals.grpc.pb.cc.obj
[ 45%] Building CXX object validation/sil/CMakeFiles/sil.dir/sil_client.cc.obj
[ 54%] Linking CXX static library libsil.a
[ 54%] Built target sil
[ 63%] Building CXX object CMakeFiles/bindings.dir/projects/Demo/CAN/Bar/platforms/raspi/bindings.cc.obj
In file included from C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/projects/Demo/CAN/Bar/platforms/raspi/bindings.cc:7:
C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/mcal/raspi/periph/can.h:6:10: fatal error: linux/can.h: No such file or directory
    6 | #include <linux/can.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
make[3]: *** [CMakeFiles/bindings.dir/build.make:77: CMakeFiles/bindings.dir/projects/Demo/CAN/Bar/platforms/raspi/bindings.cc.obj] Error 1
make[2]: *** [CMakeFiles/Makefile2:233: CMakeFiles/bindings.dir/all] Error 2
make[1]: *** [Makefile:91: all] Error 2
make[1]: Leaving directory 'C:/Users/teghv/Documents/School/MAC-FE/racecar/firmware/build/Demo/CAN/Bar/raspi'
make: *** [Makefile:23: build] Error 2
(.env)
```

</p>
</details>

--- a/scripts/build_project.sh
+++ b/scripts/build_project.sh
@@ -6,23 +6,24 @@
 # Check args
 if [ "$#" -ne 4 ]; then
     echo "Usage: $0 <commit-hash> <repo-path> <project> <platform>"
-    exit
+    exit 2
 fi
 
 # Assign inputs to variables
 COMMIT_HASH=$1
-REPO_PATH=$2
+RACECAR_PATH=$2
 PROJECT=$3
 PLATFORM=$4
 
 # Checks if directory exists
-if [ ! -d "$REPO_PATH" ]; then
-    echo "Error: Repository path '$REPO_PATH' does not exist."
-    exit
+FW_PATH=$RACECAR_PATH/firmware
+if [ ! -d "$FW_PATH" ]; then
+    echo "Error: Firmware directory not found within repository path '$RACECAR_PATH'."
+    exit 2
 fi
 
 # Navigate to the racecar/firmware directory path
-cd "$REPO_PATH"
+cd "$FW_PATH"
 
 # Fetch latest changes and check out the specified commit
 git fetch --all

--- a/scripts/build_project.sh
+++ b/scripts/build_project.sh
@@ -1,12 +1,12 @@
+#!/bin/bash
+
 # Author: Teghveer Singh Ateliey
 # Date: 2024-10-30
-
-#!/bin/bash
 
 # Check args
 if [ "$#" -ne 4 ]; then
     echo "Usage: $0 <commit-hash> <repo-path> <project> <platform>"
-    exit 1
+    exit
 fi
 
 # Assign inputs to variables
@@ -22,21 +22,17 @@ if [ ! -d "$REPO_PATH" ]; then
 fi
 
 # Navigate to the racecar/firmware directory path
-cd "$REPO_PATH" || exit
-
-# Activate the Python virtual environment
-source .env/Scripts/activate
+cd "$REPO_PATH"
 
 # Fetch latest changes and check out the specified commit
 git fetch --all
 git checkout "$COMMIT_HASH"
 
 # Build command
-if make PROJECT="$PROJECT" PLATFORM="$PLATFORM" build; then
+if make PROJECT="$PROJECT" PLATFORM="$PLATFORM" clean build; then
     echo "Build completed successfully!!"
 else
     echo "Build failed :("
-    exit
 fi
 
 # Return to the main branch

--- a/scripts/build_project.sh
+++ b/scripts/build_project.sh
@@ -40,5 +40,6 @@ else
     echo "Build completed successfully!!"
 fi
 
-# Return to the main branch
+# Return to the main branch and return make exit code
 git checkout main
+exit $MAKE_ERR_CODE

--- a/scripts/build_project.sh
+++ b/scripts/build_project.sh
@@ -1,0 +1,43 @@
+# Author: Teghveer Singh Ateliey
+# Date: 2024-10-30
+
+#!/bin/bash
+
+# Check args
+if [ "$#" -ne 4 ]; then
+    echo "Usage: $0 <commit-hash> <repo-path> <project> <platform>"
+    exit 1
+fi
+
+# Assign inputs to variables
+COMMIT_HASH=$1
+REPO_PATH=$2
+PROJECT=$3
+PLATFORM=$4
+
+# Checks if directory exists
+if [ ! -d "$REPO_PATH" ]; then
+    echo "Error: Repository path '$REPO_PATH' does not exist."
+    exit
+fi
+
+# Navigate to the racecar/firmware directory path
+cd "$REPO_PATH" || exit
+
+# Activate the Python virtual environment
+source .env/Scripts/activate
+
+# Fetch latest changes and check out the specified commit
+git fetch --all
+git checkout "$COMMIT_HASH"
+
+# Build command
+if make PROJECT="$PROJECT" PLATFORM="$PLATFORM" build; then
+    echo "Build completed successfully!!"
+else
+    echo "Build failed :("
+    exit
+fi
+
+# Return to the main branch
+git checkout main

--- a/scripts/build_project.sh
+++ b/scripts/build_project.sh
@@ -30,10 +30,14 @@ git fetch --all
 git checkout "$COMMIT_HASH"
 
 # Build command
-if make PROJECT="$PROJECT" PLATFORM="$PLATFORM" clean build; then
-    echo "Build completed successfully!!"
+make PROJECT="$PROJECT" PLATFORM="$PLATFORM" clean build
+MAKE_ERR_CODE=$?
+
+# Prints error code of make if command fails
+if [ "$MAKE_ERR_CODE" -ne 0 ]; then
+    echo "Build failed with makefile error code: $MAKE_ERR_CODE"
 else
-    echo "Build failed :("
+    echo "Build completed successfully!!"
 fi
 
 # Return to the main branch

